GoToLine: Screenreader output line number, handles #66245, #65219

cleidigh opened the pull request:
@isidorn 
@alexandrudima 

GoToLine: Screenreader output line number, handles #66245, #65219
- Add "Current Line: x" as prefix to quick open prompt and ARIA label
- Fix labels for non- editor/no editor cases

Notes:

- QuickOpen Label: determine if it should always display what looks like
  the error condition to choose a number between 1 and y

@zersiax - your specific approach would require using a different form of
the quick open as far as I can tell since the current form does not offer
the option for placeholder text/highlighted input (but perhaps I'm wrong on this).

Since the existing label output would always be there, I'm not sure the "Current Line"
it is much extra above the other standard stuff.  Either way I follow the team

```ts

	// TODO: choose output string
	// Output verbose current line / go to line label
	return nls.localize('gotoLineLabelEmpty', "Current Line: {0}. Type a line number to navigate to", line);

	// Output current line # only for prefix / go to line label (@zersiax approach)
	// return nls.localize('gotoLineLabelEmpty2', "{0}. Type a line number to navigate to", line);

```

isidorn commented on src/vs/workbench/parts/quickopen/browser/gotoLineHandler.ts:
```diff
@@ -213,7 +215,31 @@ export class GotoLineHandler extends QuickOpenHandler {
+
+		// Check we can run - (is text editor)
+		const canRun = this.canRun();
+		if (types.isUndefinedOrNull(canRun) || (typeof canRun === 'boolean' && !canRun) || typeof canRun === 'string') {
```
Why do you check for `isUndefinedOrNull`, you own the `canRun` method and it either returns `true` or `string`. It feels like this if can be simplified.

cleidigh replied:
I "borrowed"  the logic from QuickOpenController, it seems to be conservative covering the three cases so I followed the same:

```ts
private handleSpecificHandler(handlerDescriptor: QuickOpenHandlerDescriptor, value: string, token: CancellationToken): Promise<void> {
		return this.resolveHandler(handlerDescriptor).then((resolvedHandler: QuickOpenHandler) => {

			// Remove handler prefix from search value
			value = value.substr(handlerDescriptor.prefix.length);

			// Return early if the handler can not run in the current environment and inform the user
			const canRun = resolvedHandler.canRun();
			if (types.isUndefinedOrNull(canRun) || (typeof canRun === 'boolean' && !canRun) || typeof canRun === 'string') {
				const placeHolderLabel = (typeof canRun === 'string') ? canRun : nls.localize('canNotRunPlaceholder', "This quick open handler can not be used in the current context");
```

I did tweak for this and your related, below.

isidorn commented on src/vs/workbench/parts/quickopen/browser/gotoLineHandler.ts:
```diff
@@ -213,7 +215,31 @@ export class GotoLineHandler extends QuickOpenHandler {
+			// return nls.localize('gotoLineLabelEmpty2', "{0}. Type a line number to navigate to", line);
+		}
+		else {
+			return label;
```
If I understand correctly label here is always `''`. Can we simplify or fix this?

cleidigh replied:
set the label to the default case

isidorn commented:
@cleidigh nice job. I have added some minor comments and it would be good if you coud tackle them.
Unassigning @bpasero since 2 reviewers is more than enough

alexandrudima commented:
@isidorn I suggest @bpasero review this, as he owns the area. I merely tested this works.

bpasero commented:
I can review once the comments are addressed from Isi.

cleidigh commented on src/vs/workbench/parts/quickopen/browser/gotoLineHandler.ts:
```diff
@@ -91,12 +91,14 @@ class GotoLineEntry extends EditorQuickOpenEntry {
 
 		// Inform user about valid range if input is invalid
 		const maxLineNumber = this.getMaxLineNumber();
+		const line = this.editorService.activeTextEditorWidget.getPosition().lineNumber;
```
My interpretation was that the command constructor calls canRun and sets error labels upfront preventing the command and get label call.

I guess the more conservative approach is to check again in getLabel ?

I added a number test on line.

cleidigh commented:
FYI

I removed the bare-bones label option given no comments...

cleidigh commented:
thanks reviewers...

isidorn commented:
@bpasero I actually like the behavior you proposed. More simple and goes in line with some of our other designs.

bpasero commented:
@cleidigh I think one advantage of having the current line highlighted in the input is that it is very easy for the user to see what the current line is. But maybe we should really check with @zersiax first to learn what the best solution would be.

zersiax commented:
@bpasero  Having the current line highlighted or not doesn't make a huge difference; it's mainly a suggestion because a lot of edit fields with prefilled content work that way, so you can immediately type your alternative without having to delete the stuff that's there first. This also eliminates the need for an extra aria-label because the sr would essentially just read out the number in the edit field, foregoing a label altogether. But I don't know how feasible that is with the implementation of that dialogue. The output would be something like:
"Type a line number to navigate to: combobox edit field has autoComplete 54 selected,   type ? to get help to get help on the actions from here"

As you can see that message is already pretty long even without the line number. So The most ideal I think we would have, if we can't prefill the field in an sr-aware manner, is to have the current line number appended to the initial phrase. Something like:
"Type the line number to navigate to, currently 54" yadda yadda

The problem with that approach is that an SR-only user won't know there's already a line number filled in, which would require the field be cleared when the user starts typing.

bpasero commented:
> The problem with that approach is that an SR-only user won't know there's already a line number filled in, which would require the field be cleared when the user starts typing.

That would happen automatically because the line number is already pre-selected, right?

So it seems to me this approach is feasable if we can tweak the ARIA label for the input field?

zersiax commented:
@bpasero  that would depend. Are you positive the screen reader wouldn't just read the value that is prefilled in and preselected as is? If it does, we don't need to do anything with the aria label I would think, as it would just work?
If you're sure that doesn't work though, than I would say yes, you're right. Tweak the aria-label and make the line number highlighted in that case so typing blanks it and inserts the typed value.

bpasero commented:
If you configure `"workbench.commandPalette.preserveInput": true` and bring up command picker (`Ctrl+Shift+P`) you can actually simulate the behaviour. Any command you executed before will be pre-selected.

Testing that I now realise that this will not work nicely because of the `aria-descendent` property we use in the picker: Basically we always read out the selected element in the picker and never the input field. 

Still, since in the case of the "Go to Line" command we only ever have one result that is read by the screen reader we can probably make it work by just having a custom label for the very first time you open the picker (something like "Back to line 42" and once you type it changes to  "Go to line 45"). We can tweak the label.

I see how this could mean more implementation work in the end but that should not be the criteria here for finding the best solution.

zersiax commented:
ahh. Yes, if the input field doesn't get reado ut, which makes sense if its the same control as the shift+ctrl+p behavior, than we'll need to tweak the label indeed.

cleidigh commented:
All:

I partially implemented 3 more approaches I mentioned above that pretty much
cover the variety of methods floated by everyone.  I have screenshots below for anyone not
building the actual code.  no validation phase yet.  Latest Commit is dirty WIP !

Example 1:
- Current PR: based on existing QuickOpenAction method, picker prefix, cannot not have
  selected input
- ARIA label fully independent 
- some extra output possibly removable from quick open action

Example 2:
- Utilize QuickOpen.input method , input label as existing, set input value to correct line number
- add some extra uncontrollable output 'edit' and 'selected', no placeholder

Example 3:
- same as above, but add placeholder that is read at the beginning, not shown

Example 4:
- same as example 3, but no preset input or selected text, visible placeholder
- nothing to overwrite
- no prefix that can be deleted moving into file mode


1 - Current:
![image](https://user-images.githubusercontent.com/25272315/51069627-e340da00-1600-11e9-8567-089b623538c6.png)


2 
![image](https://user-images.githubusercontent.com/25272315/51069637-17b49600-1601-11e9-92b3-cb37e0bd1d3d.png)


3
![image](https://user-images.githubusercontent.com/25272315/51069650-6a8e4d80-1601-11e9-8ae6-f72447400d17.png)

4
![image](https://user-images.githubusercontent.com/25272315/51069657-7ed24a80-1601-11e9-9801-59bf50d736b6.png)

bpasero commented:
@cleidigh the "placeholder" approach would not work though because "`:`" is always in the input field, no? That is the input to actually trigger the "Go to Line" picker.

I like the approach where visually things stay the same but for screen reader we read out some extra information. Which change should I be looking at that implements 2 or 3?

cleidigh commented:
@bpasero 
Only Example 1 (GoToLineAction) utilizes the existing quick pick method.  As you stated we have the prefix but cannot use a placeholder.  That is why I experimented with the quick pick input method for examples 2-4.  I am just getting more familiar with the quick pick options so ignore any comments due to my lack of understanding.  Since the user never picks from choices it seems the input method is more appropriate.  I also think (and @zersiax confirmed) having the current line number selected text in the input really does not bias anything.  What it does do is usually had extra screenreader output.

Anyway.  I added a few comments so you can select before compile Examples 2-4 within GotoLineAction2. Note: these examples are still incomplete until we decide what to do.  Only 2-4 can have selected input without component changes below as far as I can tell.  Example 1 still uses the quick open picker and therefore the validation and execution still works.  

If you decide that the preselected input for the current line is NOT preferable, Example 1 will operate most similar to how it is now.  Example 4 is the cleanest both visually and in terms of the least superfluous output IMHO.

From GoToLineAction2()

```ts


		const line = this.editorService.activeTextEditorWidget.getPosition().lineNumber;
		const maxLine = (<ITextModel>this.editorService.activeTextEditorWidget.getModel()).getLineCount();

		return this.quickInputService.input({
			// Example 2: Populate input value, gets selected - SR outputs "xx selected", no placeholder
			value: line.toString(),
			prompt: nls.localize('gotoLine.prompt1', "Type a line number between 1 and {0} to navigate to.", maxLine),

			// Example 3: Populate input value, gets selected, set placeholder - gets read but not shown
			// value: line.toString(),
			// prompt: nls.localize('gotoLine.prompt1', "Type a line number between 1 and {0} to navigate to.", maxLine),
			// placeHolder: nls.localize('gotoLine3', "Current line: {0}.", line.toString())

			// Example 4: NO preset input value, elected, set placeholder - gets read but not shown
			// prompt: nls.localize('gotoLine.prompt1', "Type a line number between 1 and {0} to navigate to.", maxLine),
			// placeHolder: nls.localize('gotoLine3', "Current line: {0}.", line.toString())


		}).then(line => {
			if (line) {

				// need to add 	validation and error prompt - actual jump

				if (restoreOptions) {
					Event.once(this._quickOpenService.onHide)(() => {
						activeTextEditorWidget.updateOptions(restoreOptions);
					});
				}

			}
		});
```

zersiax commented:
@bpasero  Are you sure you want nothing to change visually? I believe @isidorn  expressed interest in seeing the line number as well since he has them disabled and this would be a quick way to see what line number they're on. Or am I misunderstanding either of you?

bpasero commented:
@cleidigh you say "having the current line number selected text in the input really does not add anything for the SR user" and I disagree. The point is that it helps users that have line numbers disabled to know what the current line number is, would you agree?

@zersiax the current line number is pre-filled into the input box and selected, is that not sufficient?

Anyhow, to keep this simple (since the pieces of being able to pre-select in quick open do not exist, I suggest we simply go with an approach that changes the initial message that you see:

"Current Line: {0}. Type a line number to navigate to..."

That should solve the issue for people that need a SR because we announce the text (as far as I know) when the picker opens and it allows @isidorn to know on which line number he is.

zersiax commented:
LGTM. I thought for a moment that we'd only add it to the sr label, but this approach sounds fine to me

cleidigh commented:
@bpasero 
@zersiax 
I absolutely agree having the current line number for non-SR users IS useful.  I was not clear enough in my statement.  What I meant that selected text in the input when it was  already  added to the prompt label or my other proposal in the placeholder would be redundant.

I restored the code to have line number two the prompt that also gets read.

I think this looks and sounds good, thanks all for going through the discussion.
Should be good to go.  @isidorn just want to make sure I addressed all your review comments adequately...

bpasero commented on src/vs/workbench/parts/quickopen/browser/gotoLineHandler.ts:
```diff
@@ -91,12 +91,14 @@ class GotoLineEntry extends EditorQuickOpenEntry {
+		if (this.invalidRange(maxLineNumber) && typeof line === 'number') {
 			if (maxLineNumber > 0) {
-				return nls.localize('gotoLineLabelEmptyWithLimit', "Type a line number between 1 and {0} to navigate to", maxLineNumber);
+				return nls.localize('gotoLineLabelEmptyWithLimit', "Current Line: {0}.  Type a line number between 1 and {1} to navigate to", line, maxLineNumber);
```
@cleidigh there is an extra whitespace in the localize value

cleidigh replied:
👍

bpasero commented on src/vs/workbench/parts/quickopen/browser/gotoLineHandler.ts:
```diff
@@ -91,12 +91,14 @@ class GotoLineEntry extends EditorQuickOpenEntry {
 		// Inform user about valid range if input is invalid
 		const maxLineNumber = this.getMaxLineNumber();
-		if (this.invalidRange(maxLineNumber)) {
+		const line = this.editorService.activeTextEditorWidget.getPosition().lineNumber;
```
@cleidigh I suggest to move this into the body below, e.g.:

```ts
getLabel(): string {

	// Inform user about valid range if input is invalid
	const maxLineNumber = this.getMaxLineNumber();
	if (this.invalidRange(maxLineNumber)) {
		const currentLine = this.editorService.activeTextEditorWidget.getPosition().lineNumber;

		if (maxLineNumber > 0) {
			return nls.localize('gotoLineLabelEmptyWithLimit', "Current Line: {0}. Type a line number between 1 and {1} to navigate to.", currentLine, maxLineNumber);
		}

		return nls.localize('gotoLineLabelEmpty', "Current Line: {0}. Type a line number to navigate to.", currentLine);
	}

	// Input valid, indicate action
	return this.column ? nls.localize('gotoLineColumnLabel', "Go to line {0} and character {1}", this.line, this.column) : nls.localize('gotoLineLabel', "Go to line {0}", this.line);
}
```

cleidigh replied:
Education:
Is the advantage that the interpreter frees up the local block 'line' var at the close of the block?

cleidigh replied:
👍

bpasero commented on src/vs/workbench/parts/quickopen/browser/gotoLineHandler.ts:
```diff
@@ -213,7 +215,24 @@ export class GotoLineHandler extends QuickOpenHandler {
 
 	getAriaLabel(): string {
-		return nls.localize('gotoLineHandlerAriaLabel', "Type a line number to navigate to.");
+
```
@cleidigh this is horribly complex to read and understand, why not just:

```ts
getAriaLabel(): string {
	if (this.editorService.activeTextEditorWidget) {
		const currentLine = this.editorService.activeTextEditorWidget.getPosition().lineNumber;

		return nls.localize('gotoLineAriaLabel', "Current Line: {0}. Type a line number to navigate to.", currentLine);
	}

	return nls.localize('gotoLineHandlerAriaLabel', "Type a line number to navigate to.");
}
```

cleidigh replied:
Yeah some lack of understanding between the two quickopen approaches.  

I did not realize canRun() was already run as a base method.

For fixing #66249 I found while working on this , we need to get the aria label for the non- text editor conditions.  I have brought the same label into getAriaLabel() for that which was previously done by my extraneous canRun.

I think this is the correct logic, but I am far from understanding all of the quick open logic flow.

```ts
	getAriaLabel(): string {

		if (this.editorService.activeTextEditorWidget) {
			const currentLine = this.editorService.activeTextEditorWidget.getPosition().lineNumber;
			// Output verbose current line / go to line label
			return nls.localize('gotoLineLabelEmpty', "Current Line: {0}. Type a line number to navigate to", currentLine);
		}

		return  nls.localize('cannotRunGotoLine', "Open a text file first to go to a line");
	}
```

bpasero commented on src/vs/workbench/parts/quickopen/browser/gotoLineHandler.ts:
```diff
@@ -213,7 +215,24 @@ export class GotoLineHandler extends QuickOpenHandler {
+
+		if (typeof line === 'number') {
+			// Output verbose current line / go to line label
+			label = nls.localize('gotoLineLabelEmpty', "Current Line: {0}. Type a line number to navigate to", line) + '.';
```
@cleidigh you cannot modify a string outside of the nls method, otherwise this would not be translated

cleidigh replied:
👍

bpasero commented:
@cleidigh there is unfortunately also the standalone editor version that would benefit from this change: https://github.com/Microsoft/vscode/blob/d3024ec/src/vs/editor/standalone/browser/quickOpen/gotoLine.ts#L89

cleidigh commented:
@bpasero 
I just went ahead and did the commit for the standalone editor, but I don't know how to test.  Also don't know if you want to add the current line to the visible prompts as well?

bpasero commented:
LGTM

cleidigh commented:
thanks..
